hikari/impl/voice.py

```diff
@@ -63,16 +63,46 @@ def __init__(self, app: traits.GatewayBotAware) -> None:
         self._app = app
         self._connections = {}
         self.connections = types.MappingProxyType(self._connections)
-        self._app.event_manager.subscribe(voice_events.VoiceEvent, self._on_voice_event)
+        self._is_alive = False
+        self._is_closing = False
 
-    async def disconnect(self) -> None:
+    @property
+    def is_alive(self) -> bool:
+        return self._is_alive
+
+    def _check_if_alive(self) -> None:
+        if not self._is_alive:
+            raise errors.ComponentStateConflictError("Component cannot be used while it's not alive")
+
+        if self._is_closing:
+            raise errors.ComponentStateConflictError("Component cannot be used while it's closing")
+
+    async def _disconnect(self) -> None:
         if self._connections:
             _LOGGER.info("shutting down %s voice connection(s)", len(self._connections))
+            # We rely on the assumption that _on_connection_close will be called here rather than explicitly
+            # emptying self._connections.
             await asyncio.gather(*(c.disconnect() for c in self._connections.values()))
 
+    async def disconnect(self) -> None:
+        self._check_if_alive()
+        await self._disconnect()
+
     async def close(self) -> None:
-        await self.disconnect()
+        self._check_if_alive()
+        self._is_closing = True
         self._app.event_manager.unsubscribe(voice_events.VoiceEvent, self._on_voice_event)
+        await self._disconnect()
```

Comment:
Unsubscribing from the event should go after we call `_disconnect`

Reply:
Why

Reply:
Hmmm, true. Probably best to do it this way then